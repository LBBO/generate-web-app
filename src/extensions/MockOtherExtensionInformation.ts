import { Extension } from '../core/Extension'
import { ProjectMetaData } from '../core/userDialog/PerformUserDialog'
import { generateMockPackageManagerStrategy } from '../core/packageManagers/MockPackageManagerStrategy'
import { PackageManagerNames } from '../core/packageManagers/PackageManagerStrategy'
import { generateMockExtension } from './MockExtension'
import { OverrideProperties } from '../types/UtilityTypes'

export const generateMockProjectMetadata = (
  // Override should be ProjectMetaData, but the packageManagerStrategy value
  // will be passed to the mock generator. Also, all values should be optional.
  override?: Partial<
    OverrideProperties<
      ProjectMetaData,
      {
        packageManagerStrategy: Parameters<
          typeof generateMockPackageManagerStrategy
        >[0]
      }
    >
  >,
): ProjectMetaData => ({
  name: override?.name ?? 'mock-project-name',
  packageManagerStrategy: generateMockPackageManagerStrategy(
    override?.packageManagerStrategy,
  ),
  chosenPackageManager:
    override?.chosenPackageManager ?? PackageManagerNames.NPM,
})

type OtherExtensionInformation = Parameters<Extension['run']>[1]
export const generateMockOtherExtensionInformation = (
  override?: Partial<
    OverrideProperties<
      OtherExtensionInformation,
      {
        projectMetadata: Parameters<typeof generateMockProjectMetadata>[0]
      }
    >
  >,
): OtherExtensionInformation => ({
  chosenExtensions: override?.chosenExtensions ?? [
    generateMockExtension({ name: 'Mock extension 1' }),
    generateMockExtension({ name: 'Mock extension 2' }),
    generateMockExtension({ name: 'Mock extension 3' }),
  ],
  projectMetadata: generateMockProjectMetadata(override?.projectMetadata),
})
