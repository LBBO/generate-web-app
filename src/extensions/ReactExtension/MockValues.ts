export const defaultIndexTsx = `import React from 'react'
import ReactDOM from 'react-dom'
import './index.css'
import App from './App'
import reportWebVitals from './reportWebVitals'

ReactDOM.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
  document.getElementById('root'),
)

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals()
`

export const defaultAppTsx = `import React from 'react'
import logo from './logo.svg'
import './App.scss'

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <p>
          Edit <code>src/App.tsx</code> and save to reload.
        </p>
        <span>
          <span>Learn </span>
          <a
            className="App-link"
            href="https://reactjs.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            React
          </a>
          <span>, </span>
          <a
            className="App-link"
            href="https://redux.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Redux
          </a>
          <span>, </span>
          <a
            className="App-link"
            href="https://redux-toolkit.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Redux Toolkit
          </a>
          ,<span> and </span>
          <a
            className="App-link"
            href="https://react-redux.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            React Redux
          </a>
        </span>
      </header>
    </div>
  )
}

export default App
`

export const appTsxWithAdditionalComponents = `import React from 'react'
import logo from './logo.svg'
import './App.scss'

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <PreExistingComponent />
        <AnotherComponent />
        <p>
          Edit <code>src/App.tsx</code> and save to reload.
        </p>
        <span>
          <span>Learn </span>
          <a
            className="App-link"
            href="https://reactjs.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            React
          </a>
          <span>, </span>
          <a
            className="App-link"
            href="https://redux.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Redux
          </a>
          <span>, </span>
          <a
            className="App-link"
            href="https://redux-toolkit.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Redux Toolkit
          </a>
          ,<span> and </span>
          <a
            className="App-link"
            href="https://react-redux.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            React Redux
          </a>
        </span>
      </header>
    </div>
  )
}

export default App
`
