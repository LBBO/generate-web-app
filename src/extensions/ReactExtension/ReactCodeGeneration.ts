import type { AdditionalInformationForExtensions } from '../../core/Extension'
import path from 'path'
import { readFile, writeFile } from 'fs/promises'
import { getTypeScriptExtension } from '../Getters'
import { formatWithPrettier } from '../../core/FormatCode'
import { addImportToJsOrTsFile } from '../../core/CodeGeneration'

export const addComponent = async (
  srcFilePath: string,
  componentName: string,
  otherInformation: AdditionalInformationForExtensions,
  exportType: 'default' | 'item' = 'item',
): Promise<void> => {
  const pathToAppComponent = path.join(
    otherInformation.projectMetadata.rootDirectory,
    'src',
    `App.${
      getTypeScriptExtension(otherInformation.chosenExtensions) ? 't' : 'j'
    }sx`,
  )

  await addImportToJsOrTsFile(pathToAppComponent, {
    sourcePath: srcFilePath,
    importItems: exportType === 'item' ? [componentName] : undefined,
    importDefault: exportType === 'default' ? componentName : undefined,
  })

  await readFile(pathToAppComponent)
  await writeFile(pathToAppComponent, '')
}

export const surroundAppWithComponentWithoutImport = async (
  componentOpeningTag: string,
  otherInformation: AdditionalInformationForExtensions,
): Promise<void> => {
  const pathToIndexFile = path.join(
    otherInformation.projectMetadata.rootDirectory,
    'src',
    `index.${
      getTypeScriptExtension(otherInformation.chosenExtensions) ? 't' : 'j'
    }sx`,
  )
  const fileContent = (await readFile(pathToIndexFile)).toString()
  const componentName = componentOpeningTag.match(/<(.*?)\W/)?.[1]

  if (!componentName) {
    throw new Error('Malformed opening tag!')
  }

  const componentClosingTag = `</${componentName}>`
  const newFileContent = fileContent.replace(
    /<App ?\/>/g,
    `${componentOpeningTag}<App />${componentClosingTag}`,
  )
  await writeFile(
    pathToIndexFile,
    formatWithPrettier(newFileContent, pathToIndexFile),
  )
}
