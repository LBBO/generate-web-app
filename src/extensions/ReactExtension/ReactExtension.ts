import type { Extension } from '../../core/Extension'
import { ExtensionCategory } from '../../core/Extension'
import { spawn } from 'child_process'
import { PackageManagerNames } from '../../core/packageManagers/PackageManagerStrategy'
import { ESLintExtension } from '../ESLintExtension'
import { ExtensionIndexes, getTypeScriptExtension } from '../Getters'

export type ReactExtensionOptions = Record<string, never>

export const ReactExtension: Extension = {
  name: 'React',
  index: ExtensionIndexes.REACT,
  description: 'A JavaScript library for building user interfaces.',
  linkToDocumentation: new URL('https://reactjs.org/'),
  // Exclusiveness to Angular is declared in Angular plugin
  // exclusiveTo: [],
  dependsOn: [ESLintExtension],
  category: ExtensionCategory.FRONTEND_FRAMEWORK,

  run: (options, otherInformation) => {
    return new Promise((resolve, reject) => {
      const npxArgs = [
        // Do not ask if the package should be installed
        '--yes',
        '--',
        'create-react-app',
        otherInformation.projectMetadata.name,
      ]

      if (getTypeScriptExtension(otherInformation.chosenExtensions)) {
        npxArgs.push('--template', 'typescript')
      }

      // If yarn is installed, CRA uses it unless --use-npm is set.
      // Thus, if the user chose Yarn as their package manager,
      // no extra action is required.
      if (
        otherInformation.projectMetadata.chosenPackageManager ===
        PackageManagerNames.NPM
      ) {
        npxArgs.push('--use-npm')
      }

      const child_process = spawn('npx', npxArgs, {
        stdio: 'inherit',
        shell: true,
      })

      child_process.on('error', (e) => {
        console.log(process.env.PATH)
        console.error(e)
      })

      child_process.on('close', (statusCode) => {
        console.log(`CRA finished with code ${statusCode}`)

        if (statusCode === 0) {
          resolve()
        } else {
          reject()
        }
      })
    })
  },
}
