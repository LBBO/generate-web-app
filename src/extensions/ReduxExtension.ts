import type { Extension } from '../core/Extension'
import { ExtensionCategory } from '../core/Extension'
import { getReactExtension, getTypeScriptExtension } from './Getters'
import { mkdir, readFile, writeFile } from 'fs/promises'
import path from 'path'
import {
  convertTypeScriptToFormattedJavaScript,
  formatWithPrettier,
} from '../core/FormatCode'
import { addImportToJsOrTsFile } from '../core/CodeGeneration'
import {
  addComponent,
  surroundAppWithComponentWithoutImport,
} from './ReactExtension/ReactCodeGeneration'

export type ReduxExtensionOptions = Record<string, unknown>

export const ReduxExtension: Extension = {
  name: 'Redux',
  description:
    'A state management library that makes your application logic easier to test and that comes with' +
    ' useful features like undo/redo, state persistence, and "time-travel debugging".',
  linkToDocumentation: new URL('https://redux.js.org/'),
  category: ExtensionCategory.JAVASCRIPT_LIBRARY,
  run: async (options, otherInformation) => {
    const packageManager =
      otherInformation.projectMetadata.packageManagerStrategy
    const dependenciesToInstall = ['redux', '@reduxjs/toolkit']

    if (getReactExtension(otherInformation.chosenExtensions)) {
      dependenciesToInstall.push('react-redux')
    }

    await packageManager.installDependencies(dependenciesToInstall)

    if (getReactExtension(otherInformation.chosenExtensions)) {
      await mkdir(
        path.join(otherInformation.projectMetadata.rootDirectory, 'src/app'),
        {
          recursive: true,
        },
      )
      await mkdir(
        path.join(
          otherInformation.projectMetadata.rootDirectory,
          'src/features/counter',
        ),
        {
          recursive: true,
        },
      )

      // Copy files
      const files = [
        'src/app/hooks.ts',
        'src/app/store.ts',
        'src/features/counter/Counter.css',
        'src/features/counter/Counter.tsx',
        'src/features/counter/counterAPI.ts',
        'src/features/counter/counterSlice.ts',
        'src/features/counter/counterSlice.test.ts',
      ]
      const pathToTemplateFolder = path.join(
        __dirname,
        '../../fileTemplates/extensions/ReduxExtension',
      )
      const typescriptHasBeenChosen = Boolean(
        getTypeScriptExtension(otherInformation.chosenExtensions),
      )

      for (const relativeFilePath of files) {
        const sourceFilePath = path.join(pathToTemplateFolder, relativeFilePath)
        const fileContent = (await readFile(sourceFilePath)).toString()
        let targetFilePath = path.join(
          otherInformation.projectMetadata.rootDirectory,
          relativeFilePath,
        )

        let newFileContent = fileContent
        if (!typescriptHasBeenChosen && /\.tsx?$/.test(targetFilePath)) {
          newFileContent = convertTypeScriptToFormattedJavaScript(
            fileContent,
            targetFilePath,
          )
          targetFilePath = targetFilePath.replace(/\.tsx?$/, '.js')
        }

        console.log(`Creating ${targetFilePath} from ${sourceFilePath}`)
        await writeFile(
          targetFilePath,
          formatWithPrettier(newFileContent, targetFilePath),
        )
      }

      const scriptFileExtension = typescriptHasBeenChosen ? 'tsx' : 'js'

      await addImportToJsOrTsFile(
        path.join(
          otherInformation.projectMetadata.rootDirectory,
          `src/index.${scriptFileExtension}`,
        ),
        {
          sourcePath: 'react-redux',
          importItems: ['Provider'],
        },
      )
      await addImportToJsOrTsFile(
        path.join(
          otherInformation.projectMetadata.rootDirectory,
          `src/index.${scriptFileExtension}`,
        ),
        {
          sourcePath: './app/store',
          importItems: ['store'],
        },
      )

      await surroundAppWithComponentWithoutImport(
        '<Provider store={store}>',
        otherInformation,
      )
      // Add component
      await addComponent(
        './features/counter/Counter',
        'Counter',
        otherInformation,
      )
    }
  },
}
