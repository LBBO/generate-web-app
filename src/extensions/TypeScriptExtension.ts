import type { Extension } from '../core/Extension'
import { ExtensionCategory } from '../core/Extension'

export type TypeScriptExtensionOptions = {
  enableStrictMode: boolean
}

export const TypeScriptExtension: Extension = {
  name: 'TypeScript',
  description:
    'An open-source language which builds on JavaScript by adding static type definitions.',
  linkToDocumentation: new URL('https://www.typescriptlang.org/'),
  category: ExtensionCategory.JAVASCRIPT_FLAVOR,
  declareCliOptions: (program) => {
    program.option('--ts-strict-mode', 'Install TypeScript in strict mode')
    program.option('--no-ts-strict-mode')
  },
  promptOptions: async (prompt, cliOptions) => {
    const strictModeIsPreEnabled = cliOptions.tsStrictMode as
      | boolean
      | undefined

    if (strictModeIsPreEnabled === undefined) {
      const answers = await prompt<{ typescriptStrictMode: boolean }>([
        {
          name: 'typescriptStrictMode',
          type: 'confirm',
          message:
            'Would you like to enable the TypeScript strict mode? More info: https://www.typescriptlang.org/tsconfig#strict',
          default: true,
        },
      ])
      return {
        enableStrictMode: answers.typescriptStrictMode,
      }
    } else {
      return {
        enableStrictMode: strictModeIsPreEnabled,
      }
    }
  },
  canBeSkipped: (options, otherInformation) => {
    const chosenExtensionNames = otherInformation.chosenExtensions.map(
      (extension) => extension.name,
    )
    return (
      chosenExtensionNames.includes('React') ||
      chosenExtensionNames.includes('Angular')
    )
  },
  run(): Promise<void> {
    return Promise.reject(new Error('TypeScript extension is not implemented'))
  },
}
