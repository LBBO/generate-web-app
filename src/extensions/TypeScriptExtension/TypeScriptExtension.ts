import type { Extension } from '../../core/Extension'
import { ExtensionCategory } from '../../core/Extension'
import {
  ExtensionIndexes,
  getAngularExtension,
  getReactExtension,
} from '../Getters'
import { asyncRunCommand } from '../../core/Utils'
import { readFile, writeFile } from 'fs/promises'
import path from 'path'

export type TypeScriptExtensionOptions = {
  enableStrictMode: boolean
}

export const TypeScriptExtension: Extension = {
  name: 'TypeScript',
  index: ExtensionIndexes.TYPESCRIPT,
  description:
    'An open-source language which builds on JavaScript by adding static type definitions.',
  linkToDocumentation: new URL('https://www.typescriptlang.org/'),
  category: ExtensionCategory.JAVASCRIPT_FLAVOR,
  declareCliOptions: (program) => {
    program.option('--ts-strict-mode', 'Install TypeScript in strict mode')
    program.option('--no-ts-strict-mode')
  },
  promptOptions: async (prompt, cliOptions) => {
    const strictModeIsPreEnabled = cliOptions.tsStrictMode as
      | boolean
      | undefined

    if (strictModeIsPreEnabled === undefined) {
      const answers = await prompt<{ typescriptStrictMode: boolean }>([
        {
          name: 'typescriptStrictMode',
          type: 'confirm',
          message:
            'Would you like to enable the TypeScript strict mode? More info: https://www.typescriptlang.org/tsconfig#strict',
          default: true,
        },
      ])
      return {
        enableStrictMode: answers.typescriptStrictMode,
      }
    } else {
      return {
        enableStrictMode: strictModeIsPreEnabled,
      }
    }
  },
  run: async (options, otherInformation): Promise<void> => {
    if (
      !getAngularExtension(otherInformation.chosenExtensions) &&
      !getReactExtension(otherInformation.chosenExtensions)
    ) {
      // Install Typescript and generate tsconfig.json if no framework was installed.
      // Frameworks do this on their own.
      await otherInformation.projectMetadata.packageManagerStrategy.installDependencies(
        [
          {
            name: 'typescript',
            isDevDependency: true,
          },
        ],
      )

      await asyncRunCommand('npx -p typescript tsc --init', {
        cwd: otherInformation.projectMetadata.rootDirectory,
      })
    }

    const pathToTsConfig = path.join(
      otherInformation.projectMetadata.rootDirectory,
      'tsconfig.json',
    )

    if (!getAngularExtension(otherInformation.chosenExtensions) && options) {
      // Set strict mode only if angular hasn't been chosen as it does it automatically
      const tsConfigContent = (await readFile(pathToTsConfig)).toString()
      const modifiedTsConfigContent = tsConfigContent.replace(
        /"strict": (?:true,( ?)|false,)( ?)/,
        // Ensure the indentation of the comment after "strict": ... stays correct
        `"strict": ${options.enableStrictMode ? 'true,$1$2' : 'false,'}$2`,
      )
      await writeFile(pathToTsConfig, modifiedTsConfigContent)
    }
  },
}
