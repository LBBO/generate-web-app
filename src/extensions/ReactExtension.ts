import { Extension, ExtensionCategory } from '../core/Extension'
import { spawn } from 'child_process'
import { getTypeScriptExtension } from './TypeScriptExtension'
import { PackageManager } from '../core/userDialog/ChoosePackageManager'

export const ReactExtension: Extension = {
  name: 'React',
  description: 'A JavaScript library for building user interfaces.',
  linkToDocumentation: new URL('https://reactjs.org/'),
  // Exclusiveness to Angular is declared in Angular plugin
  exclusiveTo: [],
  category: ExtensionCategory.FRONTEND_FRAMEWORK,

  run: (options, otherInformation) => {
    return new Promise((resolve, reject) => {
      const npxArgs = [
        'create-react-app',
        otherInformation.projectMetadata.name,
      ]

      if (getTypeScriptExtension(otherInformation.chosenExtensions)) {
        npxArgs.push('--template', 'typescript')
      }

      // If yarn is installed, CRA uses it unless --use-npm is set.
      // Thus, if the user chose Yarn as their package manager,
      // no extra action is required.
      if (
        otherInformation.projectMetadata.chosenPackageManager ===
        PackageManager.NPM
      ) {
        npxArgs.push('--use-npm')
      }

      const child_process = spawn('npx', npxArgs, {
        stdio: 'inherit',
      })

      child_process.on('close', (statusCode) => {
        console.log(`CRA finished with code ${statusCode}`)

        if (statusCode === 0) {
          resolve()
        } else {
          reject()
        }
      })
    })
  },
}
