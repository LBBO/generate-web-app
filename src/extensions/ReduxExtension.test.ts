import type { AdditionalInformationForExtensions } from '../core/Extension'
import { generateMockOtherExtensionInformation } from './MockOtherExtensionInformation'
import { ReduxExtension } from './ReduxExtension'
import { ReactExtension } from './ReactExtension'
import { TypeScriptExtension } from './TypeScriptExtension'
import fs, { readFile } from 'fs/promises'
import path from 'path'
import * as ReactCodeGeneration from './ReactExtension/ReactCodeGeneration'
import * as GeneralCodeGeneration from '../core/CodeGeneration'

describe('run', () => {
  let otherInformation: AdditionalInformationForExtensions
  let installDependenciesMock: jest.SpyInstance
  let writeFileMock: jest.SpyInstance
  let addImportToJsOrTsFileSpy: jest.SpyInstance
  let mkDirSpy: jest.SpyInstance
  let consoleLogSpy: jest.SpyInstance

  beforeEach(() => {
    otherInformation = generateMockOtherExtensionInformation()
    // The return type doesn't show it, but this mock data generation
    // generates mocks for everything
    installDependenciesMock = otherInformation.projectMetadata
      .packageManagerStrategy.installDependencies as unknown as jest.SpyInstance
    writeFileMock = jest.spyOn(fs, 'writeFile').mockResolvedValue()
    addImportToJsOrTsFileSpy = jest
      .spyOn(GeneralCodeGeneration, 'addImportToJsOrTsFile')
      .mockResolvedValue()
    mkDirSpy = jest.spyOn(fs, 'mkdir').mockResolvedValue(undefined)
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    consoleLogSpy = jest.spyOn(console, 'log').mockImplementation(() => {})
  })

  it('should always install redux and redux toolkit', async () => {
    await ReduxExtension.run(undefined, otherInformation)

    expect(installDependenciesMock).toHaveBeenCalledTimes(1)
    expect(installDependenciesMock.mock.calls[0][0].includes('redux')).toBe(
      true,
    )
    expect(
      installDependenciesMock.mock.calls[0][0].includes('@reduxjs/toolkit'),
    ).toBe(true)
  })

  it('should NOT install react-redux if react has NOT been chosen', async () => {
    await ReduxExtension.run(undefined, otherInformation)
    expect(
      installDependenciesMock.mock.calls[0][0].includes('react-redux'),
    ).toBe(false)
  })

  describe('when chosen alongside react', () => {
    let surroundAppWithComponentWithoutImportSpy: jest.SpyInstance
    let addComponentSpy: jest.SpyInstance

    beforeEach(() => {
      otherInformation = generateMockOtherExtensionInformation({
        chosenExtensions: [ReactExtension, TypeScriptExtension],
      })
      // The return type doesn't show it, but this mock data generation
      // generates mocks for everything
      installDependenciesMock = otherInformation.projectMetadata
        .packageManagerStrategy
        .installDependencies as unknown as jest.SpyInstance

      surroundAppWithComponentWithoutImportSpy = jest
        .spyOn(ReactCodeGeneration, 'surroundAppWithComponentWithoutImport')
        .mockResolvedValue()

      addComponentSpy = jest
        .spyOn(ReactCodeGeneration, 'addComponent')
        .mockResolvedValue()
    })

    it('should install react-redux', async () => {
      await ReduxExtension.run(undefined, otherInformation)
      expect(
        installDependenciesMock.mock.calls[0][0].includes('react-redux'),
      ).toBe(true)
    })

    it('should copy all CSS files', async () => {
      await ReduxExtension.run(undefined, otherInformation)

      expect(writeFileMock).toHaveBeenCalledWith(
        path.join(
          otherInformation.projectMetadata.rootDirectory,
          'src/features/counter/Counter.css',
        ),
        expect.anything(),
      )
    })

    it('should copy all relevant template files if typescript was chosen', async () => {
      await ReduxExtension.run(undefined, otherInformation)

      const filePaths = [
        'src/app/hooks.ts',
        'src/app/store.ts',
        'src/features/counter/Counter.tsx',
        'src/features/counter/counterAPI.ts',
        'src/features/counter/counterSlice.ts',
        'src/features/counter/counterSlice.test.ts',
      ]
      const pathToTemplateFolder = path.join(
        __dirname,
        '../../fileTemplates/extensions/ReduxExtension',
      )

      for (const relativeFilePath of filePaths) {
        const fileContent = (
          await readFile(path.join(pathToTemplateFolder, relativeFilePath))
        ).toString()

        expect(writeFileMock).toHaveBeenCalledWith(
          path.join(
            otherInformation.projectMetadata.rootDirectory,
            relativeFilePath,
          ),
          fileContent,
        )
      }
    })

    it('should compile all relevant TS files to JS before copying if TS was not chosen', async () => {
      // Remove TypeScriptExtension from chosen extensions
      otherInformation.chosenExtensions.splice(
        otherInformation.chosenExtensions.indexOf(TypeScriptExtension),
        1,
      )
      await ReduxExtension.run(undefined, otherInformation)

      const filePaths = [
        'src/app/hooks.js',
        'src/app/store.js',
        'src/features/counter/Counter.js',
        'src/features/counter/counterAPI.js',
        'src/features/counter/counterSlice.js',
        'src/features/counter/counterSlice.test.js',
      ]

      for (let i = 0; i < filePaths.length; i++) {
        const relativeFilePath = filePaths[i]
        const targetFilePath = path.join(
          otherInformation.projectMetadata.rootDirectory,
          relativeFilePath,
        )

        expect(writeFileMock).toHaveBeenCalledWith(
          targetFilePath,
          expect.anything(),
        )
        const correspondingCall = writeFileMock.mock.calls.find(
          (call) => call[0] === targetFilePath,
        )
        expect(correspondingCall).toMatchSnapshot()
      }
    })

    it('should add the state provider to the index.j|tsx', async () => {
      await ReduxExtension.run(undefined, otherInformation)

      expect(addImportToJsOrTsFileSpy).toHaveBeenCalledTimes(2)
      expect(surroundAppWithComponentWithoutImportSpy).toHaveBeenCalledTimes(1)
    })

    it('should add the component to the App.j|tsx', async () => {
      await ReduxExtension.run(undefined, otherInformation)

      expect(addComponentSpy).toHaveBeenCalledTimes(1)
    })

    it.todo(
      'should create the src/app and src/features/counter folders recursively',
    )
  })
})
