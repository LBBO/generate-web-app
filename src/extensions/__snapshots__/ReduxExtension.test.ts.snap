// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run when chosen alongside react should compile all relevant TS files to JS before copying if TS was not chosen 1`] = `
Array [
  "/home/mdk/git-clones/GitHub/generate-web-app/mock-project-target/src/app/hooks.js",
  "import { useDispatch, useSelector } from 'react-redux'

export const useAppDispatch = () => useDispatch()
export const useAppSelector = useSelector
",
]
`;

exports[`run when chosen alongside react should compile all relevant TS files to JS before copying if TS was not chosen 2`] = `
Array [
  "/home/mdk/git-clones/GitHub/generate-web-app/mock-project-target/src/app/store.js",
  "import { configureStore } from '@reduxjs/toolkit'
import { counterReducer } from '../features/counter/counterSlice'

export const store = configureStore({
  reducer: {
    counter: counterReducer,
  },
})
",
]
`;

exports[`run when chosen alongside react should compile all relevant TS files to JS before copying if TS was not chosen 3`] = `
Array [
  "/home/mdk/git-clones/GitHub/generate-web-app/mock-project-target/src/features/counter/Counter.js",
  "import React, { useState } from 'react'
import { useAppDispatch, useAppSelector } from '../../app/hooks'
import {
  decrement,
  increment,
  incrementAsync,
  incrementByAmount,
  incrementIfOdd,
  selectCount,
} from './counterSlice'
import './Counter.css'

export const Counter = () => {
  const count = useAppSelector(selectCount)
  const dispatch = useAppDispatch()
  const [incrementAmount, setIncrementAmount] = useState('2')

  const incrementValue = Number(incrementAmount) ?? 0

  return (
    <div>
      <div className={'row'}>
        <button
          className={'button'}
          aria-label=\\"Decrement value\\"
          onClick={() => dispatch(decrement())}
        >
          -
        </button>
        <span className={'value'}>{count}</span>
        <button
          className={'button'}
          aria-label=\\"Increment value\\"
          onClick={() => dispatch(increment())}
        >
          +
        </button>
      </div>
      <div className={'row'}>
        <input
          className={'textbox'}
          aria-label=\\"Set increment amount\\"
          value={incrementAmount}
          onChange={(e) => setIncrementAmount(e.target.value)}
        />
        <button
          className={'button'}
          onClick={() => dispatch(incrementByAmount(incrementValue))}
        >
          Add Amount
        </button>
        <button
          className={'asyncButton'}
          onClick={() => dispatch(incrementAsync(incrementValue))}
        >
          Add Async
        </button>
        <button
          className={'button'}
          onClick={() => dispatch(incrementIfOdd(incrementValue))}
        >
          Add If Odd
        </button>
      </div>
    </div>
  )
}
",
]
`;

exports[`run when chosen alongside react should compile all relevant TS files to JS before copying if TS was not chosen 4`] = `
Array [
  "/home/mdk/git-clones/GitHub/generate-web-app/mock-project-target/src/features/counter/counterAPI.js",
  "// A mock function to mimic making an async request for data
export const fetchCount = (amount = 1) =>
  new Promise((resolve) => setTimeout(() => resolve({ data: amount }), 500))
",
]
`;

exports[`run when chosen alongside react should compile all relevant TS files to JS before copying if TS was not chosen 5`] = `
Array [
  "/home/mdk/git-clones/GitHub/generate-web-app/mock-project-target/src/features/counter/counterSlice.js",
  "import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { fetchCount } from './counterAPI'

const initialState = {
  value: 0,
  status: 'idle',
}

// The function below is called a thunk and allows us to perform async logic. It
// can be dispatched like a regular action: \`dispatch(incrementAsync(10))\`. This
// will call the thunk with the \`dispatch\` function as the first argument. Async
// code can then be executed and other actions can be dispatched. Thunks are
// typically used to make async requests.
export const incrementAsync = createAsyncThunk(
  'counter/fetchCount',
  async (amount) => {
    const response = await fetchCount(amount)
    // The value we return becomes the \`fulfilled\` action payload
    return response.data
  },
)

export const counterSlice = createSlice({
  name: 'counter',
  initialState,
  // The \`reducers\` field lets us define reducers and generate associated actions
  reducers: {
    increment: (state) => {
      // Redux Toolkit allows us to write \\"mutating\\" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a \\"draft state\\" and produces a brand new
      // immutable state based off those changes
      state.value += 1
    },
    decrement: (state) => {
      state.value -= 1
    },
    // Use the PayloadAction type to declare the contents of \`action.payload\`
    incrementByAmount: (state, action) => {
      state.value += action.payload
    },
  },
  // The \`extraReducers\` field lets the slice handle actions defined elsewhere,
  // including actions generated by createAsyncThunk or in other slices.
  extraReducers: (builder) => {
    builder
      .addCase(incrementAsync.pending, (state) => {
        state.status = 'loading'
      })
      .addCase(incrementAsync.fulfilled, (state, action) => {
        state.status = 'idle'
        state.value += action.payload
      })
  },
})

export const { increment, decrement, incrementByAmount } = counterSlice.actions

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: \`useSelector((state: RootState) => state.counter.value)\`
export const selectCount = (state) => state.counter.value

export const incrementIfOdd = (amount) => (dispatch, getState) => {
  const currentValue = selectCount(getState())
  if (currentValue % 2 === 1) {
    dispatch(incrementByAmount(amount))
  }
}

export const counterReducer = counterSlice.reducer
",
]
`;

exports[`run when chosen alongside react should compile all relevant TS files to JS before copying if TS was not chosen 6`] = `
Array [
  "/home/mdk/git-clones/GitHub/generate-web-app/mock-project-target/src/features/counter/counterSlice.test.js",
  "import {
  counterReducer,
  decrement,
  increment,
  incrementByAmount,
} from './counterSlice'

describe('counter reducer', () => {
  const initialState = {
    value: 3,
    status: 'idle',
  }

  it('should handle initial state', () => {
    expect(counterReducer(undefined, { type: 'unknown' })).toEqual({
      value: 0,
      status: 'idle',
    })
  })

  it('should handle increment', () => {
    const actual = counterReducer(initialState, increment())
    expect(actual.value).toEqual(4)
  })

  it('should handle decrement', () => {
    const actual = counterReducer(initialState, decrement())
    expect(actual.value).toEqual(2)
  })

  it('should handle incrementByAmount', () => {
    const actual = counterReducer(initialState, incrementByAmount(2))
    expect(actual.value).toEqual(5)
  })
})
",
]
`;
